oml/losses/arcface.py

```diff
@@ -0,0 +1,133 @@
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+import torch
+from torch import nn
+from torch.nn import functional as F
+from torchvision.ops import MLP
+
+from oml.utils.misc_torch import label_smoothing
+
+
+class ArcFaceLoss(nn.Module):
+    """
+    ArcFace loss from `paper <https://arxiv.org/abs/1801.07698>`_ with possibility to use label smoothing.
+    It contains projection (num_features x num_classes) inside itself so you don't have to produce output of
+    ``num_classes`` yourself. Please make sure that class labels started with 0 and ended as ``num_classes - 1``.
+    """
+
+    criterion_name = "arcface"  # for better logging
+
+    def __init__(
+        self,
+        in_features: int,
+        num_classes: int,
+        label2category: Optional[Dict[Any, Any]] = None,
+        label_smoothing: Optional[float] = None,
+        m: float = 0.5,
+        s: float = 64,
+    ):
+        """
+        Args:
+            in_features: Input feature size
+            num_classes: Number of classes in train set
+            label2category: Optional, for label smoothing. If you will not provide it label smoothing will be global and
```

Comment:
1) let's swap `label_smoothing` and `label2category` arguments
2) `label_smoothing` -> `label_smooting_prob`
3) change the default value for it from `None` to 0
4) let's change the comment for `label2category`:
`Mapping from label to its category. If provided, label smoothing will redistribute ``label_smoothing_prob`` only inside the category corresponding to the sample's ground truth label.`

Reply:
5. Add `reduction` argument as pass it to CE